### PR TITLE
Added tooltips to roundend traitor purchase logs

### PR DESCRIPTION
**Pull request by Cruix**
untitled
You cannot see my mouse, but it is there over the image of the goggles.

Incidentally also changed how uplink_items work a bit.

Random traitor items no longer have their own separate sale items, but can select sale items from the uplink they are purchased on. If they select an item with limited stock, the stock of that item in the uplink will go down as if you had purchased it normally.

Surplus crates no longer get items on sale.

Stock of non-sale, limited stock items is no longer globally shared between all items. At the moment the only non-sale limited stock items are the magic eight ball and the KA pressure mod, which is probably why no one noticed that everyone was sharing them.

If you buy multiple of the same item, a little x2/x3/etc. will appear next to the icon. Items bought on sale or from surplus crates are considered different items for this purpose.

:cl:
add: Roundend traitor purchase logs now have tooltips showing the name, cost, and description of the items.
/:cl:

[why]: # Back when I was new to this game I remember seeing the chameleon projector in the roundend traitor report and thinking, "What the heck is that thing?". Now if someone else has the same question, they can just mouse over it and find out.

### CODE REVIEW

**Comment by Jordie0608** [comment]
Pretty cool; although I feel the description could be dropped, they tend to be pretty lengthy and not much help at the end of the round.

**Comment by JStheguy** [comment]
That defeats the point of the tooltips as "what the fuck is this shit" aides, the name of an item alone isn't always that helpful in giving a good description of what it does.

**Comment by FantasticFwoosh** [comment]
True @JStheguy but often neither does the description in being concise enough, a seperate tooltip descriptor field outside of the PDA description could work better.

**Comment by ExcessiveUseOfCobblestone** [comment]
Then have a separate PR which changes the descriptions to be more concise?

**Comment by FantasticFwoosh** [comment]
Could do @ExcessiveUseOfCobblestone but people are still going to fawn over the loss of lore.

File: code/modules/uplink/uplink_items.dm [hunk 1]
```diff
@@ -37,29 +37,29 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
 
 			if(!filtered_uplink_items[category])
 				filtered_uplink_items[category] = list()
-			filtered_uplink_items[category][item] = I
+			filtered_uplink_items[category][item] = new I.type()
```
Comment (Cyberboss) [inline code/modules/uplink/uplink_items.dm]: Why is this necessary?

File: code/modules/uplink/uplink_items.dm [hunk 2]
```diff
@@ -1384,12 +1387,11 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
...
 				continue
+			if(I.limited_stock == 0)
+				continue
 			possible_items += I
 
 	if(possible_items.len)
 		var/datum/uplink_item/I = pick(possible_items)
-		U.telecrystals -= I.cost
-		U.purchase_log.total_spent += I.cost
-		SSblackbox.record_feedback("nested tally", "traitor_uplink_items_bought", 1, list("[initial(I.name)]", "[cost]"))
```
Comment (Cyberboss) [inline code/modules/uplink/uplink_items.dm]: Why remove this feedback?

File: code/modules/uplink/uplink_items.dm [hunk 3]
```diff
@@ -1362,9 +1366,8 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
 			continue
 		crate_value -= I.cost
 		var/obj/goods = new I.item(C)
-		U.purchase_log.LogPurchase(goods, I.cost)
-
-	SSblackbox.record_feedback("nested tally", "traitor_uplink_items_bought", 1, list("[initial(name)]", "[cost]"))
```
Comment (Cyberboss) [inline code/modules/uplink/uplink_items.dm]: This one too

**Comment by ExcessiveUseOfCobblestone** [comment]
@FantasticFwoosh put it in a book/wiki/etc. A full blown description about the history of a particular piece for a syndicate agent armory seems weird anywho.

File: code/modules/uplink/uplink_items.dm (hunk 1 above)
Comment (Cruix) [reply code/modules/uplink/uplink_items.dm]: So that limited_stock is not shared between items on different uplinks

File: code/modules/uplink/uplink_items.dm (hunk 2 above)
Comment (Cruix) [reply code/modules/uplink/uplink_items.dm]: It will be done in MakePurchase

File: code/modules/uplink/uplink_items.dm (hunk 3 above)
Comment (Cruix) [reply code/modules/uplink/uplink_items.dm]: This too

**Comment by Anonmare** [comment]
Does this work with the bundles?

**Comment by Cruix** [comment]
It shows that you purchased a bundle, but not which one you got.

**Comment by ShizCalev** [comment]
Mind posting a pic of what you mean by that?

**Comment by Iamgoofball** [comment]
what the fuck are all these undocumented in the changelog changes

>Random traitor items no longer have their own separate sale items, but can select sale items from the uplink they are purchased on. If they select an item with limited stock, the stock of that item in the uplink will go down as if you had purchased it normally.

Surplus crates no longer get items on sale.

Stock of non-sale, limited stock items is no longer globally shared between all items. At the moment the only non-sale limited stock items are the magic eight ball and the KA pressure mod, which is probably why no one noticed that everyone was sharing them.

If you buy multiple of the same item, a little x2/x3/etc. will appear next to the icon. Items bought on sale or from surplus crates are considered different items for this purpose.

not only is this unrelated to the tooltips but they're thrown in

wtf

**Comment by Cruix** [comment]
>Mind posting a pic of what you mean by that?

It is exactly the same as the picture that is there, but with the name, cost, and description of a syndicate bundle and a picture of the syndicate bundle box instead of the thermal imaging glasses.

>Random traitor items no longer have their own separate sale items, but can select sale items from the uplink they are purchased on. If they select an item with limited stock, the stock of that item in the uplink will go down as if you had purchased it normally.

I changed the random item so that it had less copy-pasted code, which fixed some consistency problems with it. This is a fix for a bug that almost no one probably noticed, so it does not need a changelog.

>Surplus crates no longer get items on sale.

I made surplus crates not generate new sales when they pick their items, so you actually get 50 TC worth of stuff, instead of randomly getting more sometimes. This is a fix for a bug that almost no one probably noticed, so it does not need a changelog.

>Stock of non-sale, limited stock items is no longer globally shared between all items. 

This is a fix for a bug that almost no one probably noticed, so it does not need a changelog.

>If you buy multiple of the same item, a little x2/x3/etc. will appear next to the icon. Items bought on sale or from surplus crates are considered different items for this purpose.

The number of times you bought something does not need a changelog because it will be blatantly obvious to anyone who looks at the roundend report, and I had to make items bought on sale be considered separate because the tooltips would have different description than an item that was bought normally.

**Comment by ShizCalev** [comment]
@ the bundle icon thing: Oh, that's fine then.

**Comment by optimumtact** [comment]
is this ready?

**Comment by Cruix** [comment]
Yes.
